Use gcc visibility attribute

phprus opened the pull request:
### Description 

According to [GCC symbols visibility wiki](https://gcc.gnu.org/wiki/Visibility) page as such (with ``__attribute__ ((visibility ("default")))`` exported symbols should be explicitly marked, to allow compilating with ``-fvisibility=hidden`` flag.

Previous PR: #786, #799


Fixes #779, #713

- [x] - git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests) for details)_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [x] bug fix - _change that fixes an issue_
- [ ] new feature - _change that adds functionality_
- [x] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [ ] added - _required for new features and some bug fixes_
- [x] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [ ] needs to be updated
- [x] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

### Notify the following users
_List users with `@` to send notifications_

### Other information

phprus commented:
@pavelkumbrasev, @isaevil what do you think about this PR?

isaevil commented on .github/workflows/ci.yml:
```diff
@@ -173,7 +173,8 @@ jobs:
           mkdir build && cd build
           cmake -DCMAKE_CXX_STANDARD=${{ matrix.std }} -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} \
-            -DCMAKE_CXX_COMPILER=${{ matrix.cxx_compiler }} -DCMAKE_C_COMPILER=${{ matrix.c_compiler }} -DTBB_CPF=${{ matrix.preview }} ..
+            -DCMAKE_CXX_COMPILER=${{ matrix.cxx_compiler }} -DCMAKE_C_COMPILER=${{ matrix.c_compiler }} -DTBB_CPF=${{ matrix.preview }} \
+            -DCMAKE_VISIBILITY_INLINES_HIDDEN=TRUE -DCMAKE_C_VISIBILITY_PRESET=hidden -DCMAKE_CXX_VISIBILITY_PRESET=hidden ..
```
I don't think that this is a good idea to add these visibility flags everywhere. Lets make only one-two instances of test matrix to actually set these flags.

phprus replied:
I think we need this flags on every platform for one of ``test`` and ``example`` build.
I'll change this commit after reviewing other changes.

phprus replied:
But, I think that the use of these flags for all configurations will make it easier to identify export of symbols errors in the project.

ilya-lavrenov replied:
agree with @phprus 
it's a good practice to always set visibility to hidden and explicitly export required symbols. 

But it's better to set such flags in cmake scripts, not in CI.

phprus replied:
First export issue found by this cmake options:
https://github.com/oneapi-src/oneTBB/actions/runs/5164552958/jobs/9303530791?pr=1114

Thanks @ilya-lavrenov

isaevil commented on src/tbbmalloc_proxy/proxy.cpp:
```diff
@@ -133,7 +133,7 @@ static inline void initPageSize()
 */
 
-extern "C" void *__TBB_malloc_proxy(size_t) __TBB_ALIAS_ATTR_COPY(malloc);
+extern "C" __TBB_EXPORT void *__TBB_malloc_proxy(size_t) __TBB_ALIAS_ATTR_COPY(malloc);
```
```suggestion
extern "C" TBBMALLOC_EXPORT void *__TBB_malloc_proxy(size_t) __TBB_ALIAS_ATTR_COPY(malloc);
```

isaevil commented on src/tbbmalloc_proxy/proxy.cpp:
```diff
@@ -184,23 +184,23 @@ inline void InitOrigPointers() {}
 #endif // MALLOC_UNIXLIKE_OVERLOAD_ENABLED and MALLOC_ZONE_OVERLOAD_ENABLED
 
-void *PREFIX(malloc)(ZONE_ARG size_t size) __THROW
+__TBB_EXPORT void *PREFIX(malloc)(ZONE_ARG size_t size) __THROW
```
```suggestion
TBBMALLOC_EXPORT void *PREFIX(malloc)(ZONE_ARG size_t size) __THROW
```

isaevil commented on src/tbbmalloc_proxy/proxy.cpp:
```diff
@@ -184,23 +184,23 @@ inline void InitOrigPointers() {}
 }
 
-void *PREFIX(calloc)(ZONE_ARG size_t num, size_t size) __THROW
+__TBB_EXPORT void *PREFIX(calloc)(ZONE_ARG size_t num, size_t size) __THROW
```
```suggestion
TBBMALLOC_EXPORT void *PREFIX(calloc)(ZONE_ARG size_t num, size_t size) __THROW
```

isaevil commented on src/tbbmalloc_proxy/proxy.cpp:
```diff
@@ -184,23 +184,23 @@ inline void InitOrigPointers() {}
 }
 
-void PREFIX(free)(ZONE_ARG void *object) __THROW
+__TBB_EXPORT void PREFIX(free)(ZONE_ARG void *object) __THROW
```
```suggestion
TBBMALLOC_EXPORT void PREFIX(free)(ZONE_ARG void *object) __THROW
```

isaevil commented on src/tbbmalloc_proxy/proxy.cpp:
```diff
@@ -184,23 +184,23 @@ inline void InitOrigPointers() {}
 }
 
-void *PREFIX(realloc)(ZONE_ARG void* ptr, size_t sz) __THROW
+__TBB_EXPORT void *PREFIX(realloc)(ZONE_ARG void* ptr, size_t sz) __THROW
```
```suggestion
TBBMALLOC_EXPORT void *PREFIX(realloc)(ZONE_ARG void* ptr, size_t sz) __THROW
```

isaevil commented on src/tbbmalloc_proxy/proxy.cpp:
```diff
@@ -209,13 +209,13 @@ void *PREFIX(realloc)(ZONE_ARG void* ptr, size_t sz) __THROW
    it's formally substituted by posix_memalign and deprecated,
    so we do not expect it to cause cyclic dependency with C RTL. */
-void *PREFIX(memalign)(ZONE_ARG size_t alignment, size_t size) __THROW
+__TBB_EXPORT void *PREFIX(memalign)(ZONE_ARG size_t alignment, size_t size) __THROW
```
```suggestion
TBBMALLOC_EXPORT void *PREFIX(memalign)(ZONE_ARG size_t alignment, size_t size) __THROW
```

isaevil commented on src/tbbmalloc_proxy/proxy.cpp:
```diff
@@ -209,13 +209,13 @@ void *PREFIX(realloc)(ZONE_ARG void* ptr, size_t sz) __THROW
 
 /* valloc allocates memory aligned on a page boundary */
-void *PREFIX(valloc)(ZONE_ARG size_t size) __THROW
+__TBB_EXPORT void *PREFIX(valloc)(ZONE_ARG size_t size) __THROW
```
```suggestion
TBBMALLOC_EXPORT void *PREFIX(valloc)(ZONE_ARG size_t size) __THROW
```

isaevil commented on src/tbbmalloc_proxy/proxy.cpp:
Please replace  all `__TBB_EXPORT` in tbbmalloc to `TBBMALLOC_EXPORT`

phprus replied:
@isaevil 

``TBBMALLOC_EXPORT`` does not export symbols from ``libtbbmalloc_proxy`` library.
``TBBMALLOC_EXPORT`` required ``__TBBMALLOC_BUILD`` but ``libtbbmalloc_proxy`` defined ``__TBBMALLOCPROXY_BUILD``.

First I added the new macro ``TBBMALLOCPROXY_EXPORT`` but it will only be used in ``proxy.cpp`` on UNIX.
So I replaced it with ``__TBB_EXPORT``.


With ``__TBB_EXPORT``:
```
100% tests passed, 0 tests failed out of 136
```

With ``TBBMALLOC_EXPORT``:
```
134/137 Test #134: test_malloc_atexit .......................Subprocess aborted***Exception:   0.26 sec
free(): invalid pointer

        Start 135: test_malloc_overload
135/137 Test #135: test_malloc_overload .....................***Exception: SegFault  0.22 sec
[doctest] doctest version is "2.4.11"
[doctest] run with "--help" for options
===============================================================================
/home/runner/work/oneTBB/oneTBB/test/tbbmalloc/test_malloc_overload.cpp:398:
TEST CASE:  Main set of tests

/home/runner/work/oneTBB/oneTBB/test/tbbmalloc/test_malloc_overload.cpp:398: FATAL ERROR: test case CRASHED: SIGSEGV - Segmentation violation signal

===============================================================================
[doctest] test cases: 1 | 0 passed | 1 failed | 1 skipped
[doctest] assertions: 9 | 9 passed | 0 failed |
[doctest] Status: FAILURE!

        Start 136: test_malloc_overload_disable
136/137 Test #136: test_malloc_overload_disable .............   Passed    0.00 sec
        Start 137: test_malloc_new_handler
137/137 Test #137: test_malloc_new_handler ..................   Passed    0.01 sec

99% tests passed, 2 tests failed out of 137

Total Test time (real) = 516.90 sec
Errors while running CTest

The following tests FAILED:
    134 - test_malloc_atexit (Subprocess aborted)
    135 - test_malloc_overload (SEGFAULT)
Error: Process completed with exit code 8.
```

See:
https://github.com/oneapi-src/oneTBB/actions/runs/5158539495/jobs/9292331033
https://github.com/oneapi-src/oneTBB/actions/runs/5158539495/jobs/9292331161?pr=1114

phprus replied:
@isaevil 

No.
``libtbbmalloc_proxy`` imports symbols from ``libtbbmalloc``:
https://github.com/oneapi-src/oneTBB/blob/fc184738ebb600f772517a378c3fc8a73a6d26f5/src/tbbmalloc_proxy/proxy.cpp#L43-L44


We need separate ``TBBMALLOCPROXY_EXPORT``-like macros or direct usage of ``__TBB_EXPORT`` in ``libtbbmalloc_proxy`` sources.

But ``TBBMALLOCPROXY_EXPORT`` will not be used in headers, only in cpp.
I think using ``__TBB_EXPORT`` directly makes the code more readable.

But, I can add a ``TBBMALLOCPROXY_EXPORT`` macro if needed.

isaevil replied:
> @isaevil
> 
> No. `libtbbmalloc_proxy` imports symbols from `libtbbmalloc`:
> 
> https://github.com/oneapi-src/oneTBB/blob/fc184738ebb600f772517a378c3fc8a73a6d26f5/src/tbbmalloc_proxy/proxy.cpp#L43-L44
> 
> We need separate `TBBMALLOCPROXY_EXPORT`-like macros or direct usage of `__TBB_EXPORT` in `libtbbmalloc_proxy` sources.
> 
> But `TBBMALLOCPROXY_EXPORT` will not be used in headers, only in cpp. I think using `__TBB_EXPORT` directly makes the code more readable.
> 
> But, I can add a `TBBMALLOCPROXY_EXPORT` macro if needed.

Ok.
I would prefer to make all modules aligned so I guess we need to add `TBBMALLOCPROXY_EXPORT` as well, even for one file.

phprus replied:
Done.
Added ``TBBMALLOCPROXY_EXPORT`` macro.

isaevil commented on test/tbbmalloc/test_malloc_lib_unload.cpp:
```diff
@@ -23,24 +23,28 @@
 const char *globalCallMsg = "A TBB allocator function call is resolved into wrong implementation.";
 
 #if _WIN32||_WIN64
+    #define TEST_EXPORT __declspec(dllexport)
+#else
+    #define TEST_EXPORT __TBB_EXPORT
+#endif
```
Should we move this piece of code somewhere to common header so it can be used in all tests as `TEST_EXPORT`?

phprus replied:
In different tests, the values of the ``TEST_EXPORT`` macro are different (with and without ``__declspec(dllexport)``):

```
Searching 194 files for "__TBB_EXPORT" (case sensitive)

~/Devel/oneapi-src/oneTBB/test/tbb/test_dynamic_link.cpp:
   31  #if _WIN32 || _WIN64
   32  #define TEST_EXPORT
   33  #else
   34: #define TEST_EXPORT extern "C" __TBB_EXPORT
   35  #endif /* _WIN32 || _WIN64 */
   36  

~/Devel/oneapi-src/oneTBB/test/tbbmalloc/test_malloc_atexit.cpp:
   34  #if _WIN32||_WIN64
   35  extern __declspec(dllexport)
   36  #else
   37: __TBB_EXPORT
   38  #endif
   39  bool dll_isMallocOverloaded()

~/Devel/oneapi-src/oneTBB/test/tbbmalloc/test_malloc_lib_unload.cpp:
   25  #if _WIN32||_WIN64
   26      #define TEST_EXPORT __declspec(dllexport)
   27  #else
   28:     #define TEST_EXPORT __TBB_EXPORT
   29  #endif
   30  

~/Devel/oneapi-src/oneTBB/test/tbbmalloc/test_malloc_used_by_lib.cpp:
   27  #if _WIN32||_WIN64
   28      extern __declspec(dllexport) void callDll();
   29  #else
   30:     extern __TBB_EXPORT void callDll();
   31  #endif
   32  }

4 matches across 4 files
```

I think in this PR such a generalization does not make sense.

phprus commented:
Rebased.

Removed
> Replace __TBB_EXPORT with TBBMALLOC_EXPORT. 
> Revert "Replace __TBB_EXPORT with TBBMALLOC_EXPORT."

experiment.

Move flags into cmake script (Suggested by @ilya-lavrenov).

phprus commented:
Append commit form PR #799 (issue #713).
Warning:
```
ld: warning: cannot export hidden symbol typeinfo for tbb::detail::r1::unsafe_wait from CMakeFiles/tbb.dir/exception.cpp.o
```


Error:

```
 64/136 Test  #64: test_global_control ......................Subprocess aborted***Exception:   0.10 sec
libc++abi: terminating with uncaught exception of type doctest::detail::TestFailureException
[doctest] doctest version is "2.4.11"
[doctest] run with "--help" for options
===============================================================================
/Users/runner/work/oneTBB/oneTBB/test/tbb/test_global_control.cpp:239:
TEST CASE:  prolong lifetime advanced

/Users/runner/work/oneTBB/oneTBB/test/tbb/test_global_control.cpp:140: FATAL ERROR: REQUIRE( false ) is NOT correct!
  values: REQUIRE( false )

/Users/runner/work/oneTBB/oneTBB/test/tbb/test_global_control.cpp:239: FATAL ERROR: test case CRASHED: SIGABRT - Abort (abnormal termination) signal

===============================================================================
[doctest] test cases: 3 | 2 passed | 1 failed | 1 skipped
[doctest] assertions: 6 | 5 passed | 1 failed |
[doctest] Status: FAILURE!
```

Failed test ``test_global_control``:
https://github.com/oneapi-src/oneTBB/blob/ea4e6156a2cde839fc01b21b76cd9b5743ec0cef/test/tbb/test_global_control.cpp#L126-L143

anton-potapov commented on include/oneapi/tbb/detail/_export.h:
```diff
@@ -18,27 +18,29 @@
 #define __TBB_detail__export_H
 
 #if defined(__MINGW32__)
-    #define _EXPORT __declspec(dllexport)
```
BTW can [GENERATE_EXPORT_HEADER](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html)  be used instead of hand written ?

phprus replied:
I don't know. Need experiments.
But I think this is out of scope for this PR.

ilya-lavrenov replied:
> BTW can [GENERATE_EXPORT_HEADER](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) be used instead of hand written file?

As far as we can see TBB can be built using bazel and other tools. So, having strict dependency on cmake generation of public header file is not a good idea.

phprus commented:
Rebased.

phprus commented:
Any news on this PR?

phprus commented:
ping?

isaevil commented on include/oneapi/tbb/detail/_export.h:
```diff
@@ -18,27 +18,35 @@
+
 #if __TBBBIND_BUILD
-    #define TBBBIND_EXPORT _EXPORT
+    #define TBBBIND_EXPORT __TBB_EXPORT
```
Please add `TBBBIND_EXPORT` to function `__TBB_internal_destroy_system_topology()`, otherwise `TBBBind` won't work.

phprus replied:
But...
Why was this bug not reproduced by tests in CI...

phprus commented on CMakeLists.txt:
```diff
@@ -85,6 +90,12 @@ endif()
 # ---------------------------------------------------------------------------------------------------------
+# Setup symbol visibility properties.
+
+set(CMAKE_VISIBILITY_INLINES_HIDDEN TRUE)
+set(CMAKE_C_VISIBILITY_PRESET "hidden")
+set(CMAKE_CXX_VISIBILITY_PRESET "hidden")
```
I forgot that ``cmake_policy`` doesn't apply to subprojects.

For subprojects, we need to add:
```
set(CMAKE_POLICY_DEFAULT_CMP0063 NEW)
```

phprus commented:
I think the documentation build error is out of scope for this PR:
https://github.com/oneapi-src/oneTBB/actions/runs/5587280735/jobs/10212448857?pr=1114#step:3:83

isaevil commented:
> Append commit form PR #799 (issue #713). Warning:
> 
> ```
> ld: warning: cannot export hidden symbol typeinfo for tbb::detail::r1::unsafe_wait from CMakeFiles/tbb.dir/exception.cpp.o
> ```
> 
> Error:
> 
> ```
>  64/136 Test  #64: test_global_control ......................Subprocess aborted***Exception:   0.10 sec
> libc++abi: terminating with uncaught exception of type doctest::detail::TestFailureException
> [doctest] doctest version is "2.4.11"
> [doctest] run with "--help" for options
> ===============================================================================
> /Users/runner/work/oneTBB/oneTBB/test/tbb/test_global_control.cpp:239:
> TEST CASE:  prolong lifetime advanced
> 
> /Users/runner/work/oneTBB/oneTBB/test/tbb/test_global_control.cpp:140: FATAL ERROR: REQUIRE( false ) is NOT correct!
>   values: REQUIRE( false )
> 
> /Users/runner/work/oneTBB/oneTBB/test/tbb/test_global_control.cpp:239: FATAL ERROR: test case CRASHED: SIGABRT - Abort (abnormal termination) signal
> 
> ===============================================================================
> [doctest] test cases: 3 | 2 passed | 1 failed | 1 skipped
> [doctest] assertions: 6 | 5 passed | 1 failed |
> [doctest] Status: FAILURE!
> ```
> 
> Failed test `test_global_control`:
> 
> https://github.com/oneapi-src/oneTBB/blob/ea4e6156a2cde839fc01b21b76cd9b5743ec0cef/test/tbb/test_global_control.cpp#L126-L143

@phprus Looks like the changes break ABI on MacOS, because symbols are now different:
```
$ diff patch_tbb_mac.txt vanilla_tbb_mac.txt
119c119
< __ZTIN3tbb6detail2r111unsafe_waitE S
---
> __ZTIN3tbb6detail2r111unsafe_waitE D
132c132
< __ZTVN3tbb6detail2r111unsafe_waitE S
---
> __ZTVN3tbb6detail2r111unsafe_waitE D
```
And `test_global_control` linked against `v2021.1.0` fails with the same issue you mentioned above (btw, how did you get this error?)

phprus commented:
@isaevil 

Demangled:
```
phprus@mbp tmp % c++filt __ZTIN3tbb6detail2r111unsafe_waitE
typeinfo for tbb::detail::r1::unsafe_wait

phprus@mbp tmp % c++filt __ZTVN3tbb6detail2r111unsafe_waitE
vtable for tbb::detail::r1::unsafe_wait
```


Symbols with this PR:

```
cmake -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_STANDARD=20 ../..
make -j8


# Get typeinfo for all exception:
nm -g appleclang_14.0_cxx20_64_release/libtbb.12.dylib | grep __ZTIN3tbb6detail2r
0000000000020610 S __ZTIN3tbb6detail2r110user_abortE
0000000000020590 S __ZTIN3tbb6detail2r111unsafe_waitE
0000000000020650 S __ZTIN3tbb6detail2r112missing_waitE
00000000000205d0 S __ZTIN3tbb6detail2r114bad_last_allocE


# Get vtable for all exception:
nm -g appleclang_14.0_cxx20_64_release/libtbb.12.dylib | grep __ZTVN3tbb6detail2r
00000000000205e8 S __ZTVN3tbb6detail2r110user_abortE
0000000000020568 S __ZTVN3tbb6detail2r111unsafe_waitE
0000000000020628 S __ZTVN3tbb6detail2r112missing_waitE
00000000000205a8 S __ZTVN3tbb6detail2r114bad_last_allocE
```



Symbols without this PR (current master, commit 9314fcfdaa3df8e990635abac20910922f8f0720):

```
cmake -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_STANDARD=20 ../..
make -j8

# Get typeinfo for all exception:
nm -g appleclang_14.0_cxx20_64_release/libtbb.12.dylib | grep __ZTIN3tbb6detail2r
00000000000205d0 S __ZTIN3tbb6detail2r110user_abortE
0000000000020610 S __ZTIN3tbb6detail2r112missing_waitE
0000000000020590 S __ZTIN3tbb6detail2r114bad_last_allocE

# Get vtable for all exception:
nm -g appleclang_14.0_cxx20_64_release/libtbb.12.dylib | grep __ZTVN3tbb6detail2r
00000000000205a8 S __ZTVN3tbb6detail2r110user_abortE
0000000000020640 S __ZTVN3tbb6detail2r111unsafe_waitE
00000000000205e8 S __ZTVN3tbb6detail2r112missing_waitE
0000000000020568 S __ZTVN3tbb6detail2r114bad_last_allocE
```


Without patch typeinfo for tbb::detail::r1::unsafe_wait is not exported and this exception is not captured correctly.

Clang:
```
phprus@mbp cxx20 % clang++ --version
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

For get this error build this PR without commit "Fix: ld: warning: cannot export hidden symbol typeinfo for tbb::detail::r1::unsafe_wait" (36f01083d327bd1b9ff72bc0c6ec61b1adcded44)

phprus commented:
@isaevil 

Symbols from release binary:

```
phprus@mbp oneapi-tbb-2021.9.0 % nm -g lib/libtbb.12.9.dylib | grep __ZTIN3tbb6detail2r
0000000000067b20 D __ZTIN3tbb6detail2r110user_abortE
0000000000067aa8 D __ZTIN3tbb6detail2r111unsafe_waitE
0000000000067b58 D __ZTIN3tbb6detail2r112missing_waitE
0000000000067ae0 D __ZTIN3tbb6detail2r114bad_last_allocE
phprus@mbp oneapi-tbb-2021.9.0 % nm -g lib/libtbb.12.9.dylib | grep __ZTVN3tbb6detail2r
00000000000616f0 S __ZTVN3tbb6detail2r110user_abortE
00000000000616a0 S __ZTVN3tbb6detail2r111unsafe_waitE
0000000000061718 S __ZTVN3tbb6detail2r112missing_waitE
00000000000616c8 S __ZTVN3tbb6detail2r114bad_last_allocE
```


Default build (RelWithDebInfo):
```
cmake -DCMAKE_VERBOSE_MAKEFILE=ON ../..
make -j8
```

Without PR:
```
phprus@mbp cxx % nm -g appleclang_14.0_cxx11_64_relwithdebinfo/libtbb.12.10.dylib | grep __ZTIN3tbb6detail2r
00000000000245c8 S __ZTIN3tbb6detail2r110user_abortE
0000000000024608 S __ZTIN3tbb6detail2r112missing_waitE
0000000000024588 S __ZTIN3tbb6detail2r114bad_last_allocE
phprus@mbp cxx % nm -g appleclang_14.0_cxx11_64_relwithdebinfo/libtbb.12.10.dylib | grep __ZTVN3tbb6detail2r
00000000000245a0 S __ZTVN3tbb6detail2r110user_abortE
0000000000024638 S __ZTVN3tbb6detail2r111unsafe_waitE
00000000000245e0 S __ZTVN3tbb6detail2r112missing_waitE
0000000000024560 S __ZTVN3tbb6detail2r114bad_last_allocE
```


With PR:
```
phprus@mbp cxx % nm -g appleclang_14.0_cxx11_64_relwithdebinfo/libtbb.12.10.dylib | grep __ZTIN3tbb6detail2r
0000000000024608 S __ZTIN3tbb6detail2r110user_abortE
0000000000024588 S __ZTIN3tbb6detail2r111unsafe_waitE
0000000000024648 S __ZTIN3tbb6detail2r112missing_waitE
00000000000245c8 S __ZTIN3tbb6detail2r114bad_last_allocE
phprus@mbp cxx % nm -g appleclang_14.0_cxx11_64_relwithdebinfo/libtbb.12.10.dylib | grep __ZTVN3tbb6detail2r
00000000000245e0 S __ZTVN3tbb6detail2r110user_abortE
0000000000024560 S __ZTVN3tbb6detail2r111unsafe_waitE
0000000000024620 S __ZTVN3tbb6detail2r112missing_waitE
00000000000245a0 S __ZTVN3tbb6detail2r114bad_last_allocE
```

I can't reproduce your environment and bug on local.
I don't see changing symbol types, only exporting a new symbol.

phprus commented:
@isaevil 
No, "Wihtout PR" - current master (commit 88f73bbb48d5384ccba3e35b5c5a59451f718063

phprus commented:
@isaevil , @pavelkumbrasev

Legacy macOS 10.15 (x86_64), AppleClang 12.0.0.12000032:

Commit 88f73bbb48d5384ccba3e35b5c5a59451f718063 (current master without this PR):

```
cmake -DCMAKE_VERBOSE_MAKEFILE=ON ../..
make -j8


phprus@mbp-phprus cxx % nm -g appleclang_12.0_cxx11_64_relwithdebinfo/libtbb.12.10.dylib | grep __ZTIN3tbb6detail2r
000000000002c3f0 S __ZTIN3tbb6detail2r110user_abortE
000000000002c448 S __ZTIN3tbb6detail2r111unsafe_waitE
000000000002c430 S __ZTIN3tbb6detail2r112missing_waitE
000000000002c3b0 S __ZTIN3tbb6detail2r114bad_last_allocE

phprus@mbp-phprus cxx % nm -g appleclang_12.0_cxx11_64_relwithdebinfo/libtbb.12.10.dylib | grep __ZTVN3tbb6detail2r
000000000002c3c8 S __ZTVN3tbb6detail2r110user_abortE
000000000002c460 S __ZTVN3tbb6detail2r111unsafe_waitE
000000000002c408 S __ZTVN3tbb6detail2r112missing_waitE
000000000002c388 S __ZTVN3tbb6detail2r114bad_last_allocE
```

``__ZTIN3tbb6detail2r111unsafe_waitE`` - is exported, but begin from Xcode 13.2.1 is NOT exported.
I can't check Xcode 13.2.1 on x86_64, because my legacy mac does not support this version.

As you can see, the symbols type is ``S``.

phprus commented:
My build with or without this PR:
```
phprus@mbp cxx % nm -m appleclang_14.0_cxx11_64_relwithdebinfo/libtbb.12.dylib | grep __ZTIN3tbb6detail2r112missing_waitE
0000000000024648 (__DATA_CONST,__const) external __ZTIN3tbb6detail2r112missing_waitE
```

library from binary package:
```
phprus@mbp oneapi-tbb-2021.9.0 % nm -m lib/libtbb.12.dylib|grep __ZTIN3tbb6detail2r112missing_waitE
0000000000067b58 (__DATA,__data) weak external __ZTIN3tbb6detail2r112missing_waitE
```

The class ``missing_wait`` does not change in this PR.

I have no idea why the binary package is different from the build results of the current master.

phprus commented:
Rebased.

phprus commented:
@isaevil , @pavelkumbrasev,
Can you tell me what versions of macOS and Apple Clang (XCode) are you using for compilation?

isaevil commented:
> @isaevil , @pavelkumbrasev, Can you tell me what versions of macOS and Apple Clang (XCode) are you using for compilation?

The symbols I provided earlier are from library built on MacOS 10.13 with Apple Clang 9.1.0.

phprus commented:
Reproducing:

***oneTBB-37a89d47f2bca3700f9777732c41f08ac44d2d0a***

```
MBP-phprus:build phprus$ cmake -DCMAKE_VERBOSE_MAKEFILE=ON ..
CMake Deprecation Warning at CMakeLists.txt:15 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The CXX compiler identification is AppleClang 9.1.0.9020039
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CMAKE_BUILD_TYPE is not specified. Using default: RelWithDebInfo
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Looking for getcontext
-- Looking for getcontext - found
-- IPO enabled
-- The C compiler identification is AppleClang 9.1.0.9020039
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- TBBBind build targets are disabled due to unsupported environment
-- Performing Test LINKER_HAS_NO_AS_NEEDED
-- Performing Test LINKER_HAS_NO_AS_NEEDED - Failed
-- Configuring done (10.3s)
-- Generating done (6.8s)
-- Build files have been written to: /Users/phprus/Devel/oneTBB/oneTBB-37a89d47f2bca3700f9777732c41f08ac44d2d0a/build
```

***oneTBB-unix-visibility***

```
MBP-phprus:build phprus$ cmake -DCMAKE_VERBOSE_MAKEFILE=ON ..
CMake Deprecation Warning at CMakeLists.txt:15 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The CXX compiler identification is AppleClang 9.1.0.9020039
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CMAKE_BUILD_TYPE is not specified. Using default: RelWithDebInfo
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Looking for getcontext
-- Looking for getcontext - found
-- IPO enabled
-- The C compiler identification is AppleClang 9.1.0.9020039
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- TBBBind build targets are disabled due to unsupported environment
-- Performing Test LINKER_HAS_NO_AS_NEEDED
-- Performing Test LINKER_HAS_NO_AS_NEEDED - Failed
-- Configuring done (4.3s)
-- Generating done (6.9s)
-- Build files have been written to: /Users/phprus/Devel/oneTBB/oneTBB-unix-visibility/build
```


***oneapi-tbb-2021.10.0***

From github without compilation.



Get symbols:

```
nm -g -U lib/libtbb.12.dylib | awk '{ print $2 " " $3 }' > ~/tbb-release-10.txt

nm -g -U appleclang_9.1_cxx11_64_relwithdebinfo/libtbb.12.dylib | awk '{ print $2 " " $3 }' > ~/tbb-unix-visibility.txt

nm -g -U appleclang_9.1_cxx11_64_relwithdebinfo/libtbb.12.dylib | awk '{ print $2 " " $3 }' > ~/tbb-37a89d47f2bca3700f9777732c41f08ac44d2d0a.txt
```


Check:

```
MBP-phprus:~ phprus$ md5  tbb-*
MD5 (tbb-37a89d47f2bca3700f9777732c41f08ac44d2d0a.txt) = 2002f0b1d7f1237609fd08d743ea0555
MD5 (tbb-release-10.txt) = 2002f0b1d7f1237609fd08d743ea0555
MD5 (tbb-unix-visibility.txt) = 878c9ce03ecb55967e9acae2be02a983
```


Diff:

```
diff tbb-release-10.txt tbb-unix-visibility.txt
94c94
< D __ZTIN3tbb6detail2r111unsafe_waitE
---
> S __ZTIN3tbb6detail2r111unsafe_waitE
98c98
< D __ZTVN3tbb6detail2r111unsafe_waitE
---
> S __ZTVN3tbb6detail2r111unsafe_waitE
```

phprus commented:
@isaevil, @pavelkumbrasev,


After two days of research...

The ``tbb::detail::r1::unsafe_wait`` ABI is completely broken on macOS.

I don't know how to fix it without breaking the ``tbb::detail::r1::unsafe_wait`` ABI on macOS.

Step 1:
Add visibility hidden for test ``test_tbb_header`` (commit: https://github.com/phprus/oneTBB/commit/67b2eb8cbfaffc617f3f01e42fc56f19f1e4b490).
All tests is green: https://github.com/phprus/oneTBB/actions/runs/5672150096

Step 2:
Add try-catch test for ``unsafe_wait`` (commit: https://github.com/phprus/oneTBB/commit/91aa494b072e03db4969f8db7de4d08450ca94cf).
macOS tests failed: https://github.com/phprus/oneTBB/actions/runs/5672505126

Test https://github.com/phprus/oneTBB/actions/runs/5672505126/job/15371854722
```
 71/136 Test  #71: test_tbb_header ..........................***Failed    0.10 sec
[doctest] doctest version is "2.4.11"
[doctest] run with "--help" for options
===============================================================================
/Users/runner/work/oneTBB/oneTBB/test/tbb/test_tbb_header.cpp:299:
TEST CASE:  Test for deifinition presence

/Users/runner/work/oneTBB/oneTBB/test/tbb/test_tbb_header.cpp:149: ERROR: CHECK( false ) is NOT correct!
  values: CHECK( false )
  logged: Unrecognized exception. Likely RTTI related exports are missing

===============================================================================
/Users/runner/work/oneTBB/oneTBB/test/tbb/test_tbb_header.cpp:306:
TEST CASE:  Test for multiple definition linker error

/Users/runner/work/oneTBB/oneTBB/test/tbb/test_tbb_header.cpp:149: ERROR: CHECK( false ) is NOT correct!
  values: CHECK( false )
  logged: Unrecognized exception. Likely RTTI related exports are missing

===============================================================================
[doctest] test cases:  2 |  0 passed | 2 failed | 0 skipped
[doctest] assertions: 66 | 64 passed | 2 failed |
[doctest] Status: FAILURE!
```



***ARM symbols***

Bad symbol ``tbb::detail::r1::unsafe_wait``:
```
phprus@mbp arm64-base % nm -m appleclang_14.0_cxx11_64_relwithdebinfo/libtbb.12.dylib | grep unsafe_wait | c++filt
000000000000cad0 (__TEXT,__text) non-external tbb::detail::r1::unsafe_wait::unsafe_wait(char const*)
000000000000cb00 (__TEXT,__text) non-external tbb::detail::r1::unsafe_wait::~unsafe_wait()
000000000000cafc (__TEXT,__text) non-external tbb::detail::r1::unsafe_wait::~unsafe_wait()
0000000000024620 (__DATA_CONST,__const) non-external (was a private external) typeinfo for tbb::detail::r1::unsafe_wait
0000000000020b6b (__TEXT,__const) non-external (was a private external) typeinfo name for tbb::detail::r1::unsafe_wait
0000000000024638 (__DATA_CONST,__const) weak external vtable for tbb::detail::r1::unsafe_wait
```

Good symbol ``tbb::detail::r1::user_abort``:
```
phprus@mbp arm64-base % nm -m appleclang_14.0_cxx11_64_relwithdebinfo/libtbb.12.dylib | grep user_abort | c++filt
000000000000ca10 (__TEXT,__text) non-external tbb::detail::r1::user_abort::~user_abort()
000000000000ca0c (__TEXT,__text) non-external tbb::detail::r1::user_abort::~user_abort()
000000000000c294 (__TEXT,__text) non-external tbb::detail::r1::user_abort::what() const
00000000000245c8 (__DATA_CONST,__const) external typeinfo for tbb::detail::r1::user_abort
0000000000020b2f (__TEXT,__const) non-external typeinfo name for tbb::detail::r1::user_abort
00000000000245a0 (__DATA_CONST,__const) external vtable for tbb::detail::r1::user_abort
```

***x86_64 symbols***

Bad symbol ``tbb::detail::r1::unsafe_wait``:
```
phprus@mbp x86_64-base % nm -m appleclang_14.0_cxx11_64_relwithdebinfo/libtbb.12.dylib | grep unsafe_wait | c++filt
000000000000b8f0 (__TEXT,__text) non-external tbb::detail::r1::unsafe_wait::unsafe_wait(char const*)
000000000000b930 (__TEXT,__text) non-external tbb::detail::r1::unsafe_wait::~unsafe_wait()
000000000000b920 (__TEXT,__text) non-external tbb::detail::r1::unsafe_wait::~unsafe_wait()
0000000000024698 (__DATA_CONST,__const) weak external typeinfo for tbb::detail::r1::unsafe_wait
0000000000020cf2 (__TEXT,__const) non-external (was a private external) typeinfo name for tbb::detail::r1::unsafe_wait
00000000000246b0 (__DATA_CONST,__const) weak external vtable for tbb::detail::r1::unsafe_wait
```

Good symbol ``tbb::detail::r1::user_abort``:
```
phprus@mbp x86_64-base % nm -m appleclang_14.0_cxx11_64_relwithdebinfo/libtbb.12.dylib | grep user_abort | c++filt
000000000000b7f0 (__TEXT,__text) non-external tbb::detail::r1::user_abort::~user_abort()
000000000000b7e0 (__TEXT,__text) non-external tbb::detail::r1::user_abort::~user_abort()
000000000000aff0 (__TEXT,__text) non-external tbb::detail::r1::user_abort::what() const
0000000000024640 (__DATA_CONST,__const) external typeinfo for tbb::detail::r1::user_abort
0000000000020cb6 (__TEXT,__const) non-external typeinfo name for tbb::detail::r1::user_abort
0000000000024618 (__DATA_CONST,__const) external vtable for tbb::detail::r1::user_abort
```

See symbol ``typeinfo for tbb::detail::r1::unsafe_wait``.


If this bug is not fixed, then exceptions of type ``tbb::detail::r1::unsafe_wait`` may not be caught in user code.


What do you think about it?

pavelkumbrasev commented:
@vossmjp Could you please take a look?

phprus commented:
@vossmjp,
Could you please tell what you think about the issue with ``unsafe_wait`` on macOS?

phprus commented:
Rebased.
Documentation build error is out of scope for this PR.

phprus commented:
ping?

isaevil commented:
@phprus It seems like that these two issues we are trying to solve here are orthogonal. First one about just adding `export` attributes doesn't bring any ABI breaks at first glance and can be merged, but the current second issue's resolution about `fvisibility=hidden` on application and `unsafe_wait` does bring an ABI break for existing application which is not desired. I think the second issue requires additional time and effort to investigate, but while being separated from the main content of this PR. So I would suggest doing these things:

1. Revert `tbb::unsafe_wait` changes in this PR
2. Remove setting of `-fvisilibilty=hidden` globally in root CMakelists.txt

phprus commented:
@isaevil 
I think it's better to add a cmake option to select the behavior on macOS.
What do you think?

isaevil commented:
> @isaevil I think it's better to add a cmake option to select the behavior on macOS. What do you think?

Not sure about that. What behavior do you suggest to select?

phprus commented:
@isaevil
Can you please tell me if macOS ARM ABI is stable?
If yes, which version of Xcode?
I seem to have found a workaround...

isaevil commented:
> @isaevil Can you please tell me if macOS ARM ABI is stable? If yes, which version of Xcode? I seem to have found a workaround...

I didn't check macOS ARM but I guess that it will suffer with the same issue after applying this `unsafe_wait` change because the way symbols are changed is similar to x86.

phprus commented:
@isaevil 
I found a workaround for ``tbb::detail::r1::unsafe_wait`` without breaking ABI on macOS (on all my tests).
Check ABI please.

phprus commented:
@isaevil
Any news?

phprus commented:
Rebased to resolve conflicts with current master.

phprus commented:
ping?

phprus commented:
Any news?

jdumas commented:
I admire your patience @phprus 😅 . I would also love to see this feature merged one day...

pavelkumbrasev commented:
> @phprus It seems like that these two issues we are trying to solve here are orthogonal. First one about just adding `export` attributes doesn't bring any ABI breaks at first glance and can be merged, but the current second issue's resolution about `fvisibility=hidden` on application and `unsafe_wait` does bring an ABI break for existing application which is not desired. I think the second issue requires additional time and effort to investigate, but while being separated from the main content of this PR. So I would suggest doing these things:
> 
> 1. Revert `tbb::unsafe_wait` changes in this PR
> 2. Remove setting of `-fvisilibilty=hidden` globally in root CMakelists.txt

I believe this comment is still on the table. oneTBB should preserve backward compatibility.

phprus commented:
@pavelkumbrasev

> I believe this comment is still on the table. oneTBB should preserve backward compatibility.

You are right!

On my tests, commit https://github.com/oneapi-src/oneTBB/pull/1114/commits/2887a2f49cb47d6bd9bbdcd21284dfa5e48ab4a5 (``Workaround for tbb::detail::r1::unsafe_wait broken ABI on macOS``) fix this ABI backward compatibility issue.

My previous comment about it:
https://github.com/oneapi-src/oneTBB/pull/1114#issuecomment-1689769836

Please check ABI compatibility in your environment.

pavelkumbrasev commented:
@isaevil could you please take a look ?

phprus commented:
@isaevil could you please check ABI on macOS for the current version of PR?

phprus commented:
CI errors is not related with this PR:

```
The following tests FAILED:
     33 - test_buffer_node (Timeout)
```

saschasc commented:
I have the same issue. I'm trying to build onetbb with gcc 14.x on macOS. So, I really appreciate the effort to get this in the next version.

What about the failing test? Is this known to be flaky? Can the CI be rerun?

saschasc commented:
@isaevil Anything blocking this PR for being merged beside the reviews? Would be really great if this would make it to the next version.

isaevil commented:
@saschasc I am going to validate the patch in terms of testing failures/ABI compatibility this or next week.

isaevil reviewed (approved):
I've compared ABI between current master branch and this PR on Linux and MacOS and seems like it is now the same. Workaround for `unsafe_wait` on MacOS seems to work as well.
@pavelkumbrasev what do you think?

pavelkumbrasev commented:
Since visibility attribute in a sense duplicate Def files we will need to discuss removal of Def files.

phprus commented:
> Since visibility attribute in a sense duplicate Def files we will need to discuss removal of Def files.

During the review of this PR or separately?

pavelkumbrasev commented:
I think separately, we will need to discuss it within the team.

phprus commented:
@pavelkumbrasev what do you think about this PR?

pavelkumbrasev commented:
> @pavelkumbrasev what do you think about this PR?

It looks good, we just need to define a plan for removing the Def files.

saschasc commented:
Thanks a lot for this PR and the reviews. Is it already known when the next release of oneTBB will be available. For macOS it would be great to have a fixed version as soon as possible. Thanks!

isaevil commented:
> Thanks a lot for this PR and the reviews. Is it already known when the next release of oneTBB will be available. For macOS it would be great to have a fixed version as soon as possible. Thanks!

FYI @omalyshe
